experimental/parser/parse_decl.go

```diff
@@ -0,0 +1,695 @@
+					where: taxa.MethodOuts.After(),
+					prev:  prev,
+					got:   taxa.MethodIns,
+				})
+			case !args.Options.Nil():
+				p.Error(errUnexpected{
+					what:  next,
+					where: taxa.CompactOptions.After(),
+					prev:  args.Options,
+					got:   taxa.MethodIns,
+				})
+			case !braces.Nil():
+				p.Error(errUnexpected{
+					what:  next,
+					where: taxa.Body.After(),
+					prev:  braces,
+					got:   taxa.MethodIns,
+				})
+			}
+
+			inputs = next
+			continue
+
+		case "returns": // Method outputs.
+			// Note that the inputs and outputs of a method are parsed
+			// separately, so foo(bar) and foo returns (bar) are both possible.
+			returns := c.Pop()
+
+			var ty ast.TypeAny
+			list, err := p.Punct(c, "(", taxa.KeywordReturns.After())
+			if list.Nil() && canStartPath(c.Peek()) {
+				// Suppose the user writes `returns my.Response`. This is
+				// invalid but reasonable so we want to diagnose it. To do this,
+				// we parse a single type w/o parens and diagnose it later.
+				ty, _ = parseType(p, c, taxa.KeywordReturns.After(), false)
+			} else if err != nil {
+				p.Error(err)
+				continue
+			}
+
+			var withRet report.Span
+			if !list.Nil() {
+				withRet = report.Join(returns, list)
+			} else {
+				withRet = report.Join(returns, ty)
+			}
+
+			if !outputs.Nil() || !outputTy.Nil() {
+				var out report.Spanner = outputs
+				if outputs.Nil() {
+					out = outputTy
+				}
+
+				p.Error(errMoreThanOne{
+					first:  report.Join(args.Returns, out),
+					second: withRet,
+					what:   taxa.MethodOuts,
+				})
+				continue
+			}
+
+			switch {
+			case !args.Options.Nil():
+				p.Error(errUnexpected{
+					what:  withRet,
+					where: taxa.CompactOptions.After(),
+					prev:  args.Options,
+					got:   taxa.MethodOuts,
+				})
+			case !braces.Nil():
+				p.Error(errUnexpected{
+					what:  withRet,
+					where: taxa.Body.After(),
+					prev:  braces,
+					got:   taxa.MethodOuts,
+				})
+			}
+
+			args.Returns = returns
+			if !list.Nil() {
+				outputs = list
+			} else {
+				outputTy = ty
+			}
+			continue
+
+		case "[": // Compact options.
+			c.Pop()
+			if next.IsLeaf() {
+				// Lexer has already diagnosed this for us.
+				continue
+			}
+
+			if !args.Options.Nil() {
+				p.Error(errMoreThanOne{
+					first:  args.Options,
+					second: next,
+					what:   taxa.CompactOptions,
+				})
+				continue
+			}
+
+			if !braces.Nil() {
+				p.Error(errUnexpected{
+					what:  next,
+					where: taxa.Body.After(),
+					prev:  braces,
+					got:   taxa.CompactOptions,
+				})
+			}
+
+			args.Options = parseOptions(p, next, taxa.Def)
+			continue
+
+		case "{": // Body for the definition.
+			if !braces.Nil() {
+				// We *do not* pop or diagnose an extra {}; we want this to be
+				// parsed as a loose DeclBody, instead, so we stop parsing here.
+				break followers
+			}
+
+			c.Pop()
+			if next.IsLeaf() {
+				// Lexer has already diagnosed this for us.
+				continue
+			}
+
+			braces = next
+			continue
```

Comment:
I disagree. If someone writes `oneof foo { xyz } [a=b]` or `oneof foo { xyz } = 5;`, we want to diagnose that.

Reply:
Maybe. We would diagnose it either way, just describe the error differently `expecting a declaration but instead got "="` instead of `"=" is not allowed after body`. The latter doesn't seem particularly better. And I would much rather have simpler, easier-to-read-and-maintain logic and imperfect diagnosis (which is a fact of life -- we will not be able to "perfectly" diagnose everything) than harder to maintain implementation logic but slightly better diagnoses.

I feel like we need a "complexity budget", and we should be willing to make tradeoffs to omit some of the more sophisticated diagnoses if it makes the code easier to read and comprehend. This is especially true for things of questionable frequency (we don't have any data suggesting that this is a particularly common error or that the messages produced from a more naive parser is harder for users to understand and fix, so it's unclear that tolerating more complexity to diagnose it this way is actually worth it).

Reply:
I'll just break after the first body, then. I kinda disagree re: complexity, but I don't want to hold up on the small details.